### ISSUE TITLE
Validate coverage files

### ISSUE THREAD

**Issue by Flamefire** [codecov/support#167]
The bash script that gathers coverage files is very greedy matching a lot of possible coverage files. As it is not possible to configure the matched extensions, there can be false positives. In my case I got `*.lst` files containing binary data. However the codecov.io seems to detect this as a valid file and produces wrong coverage reports (completely bogus)
So it seems there is not enough checking, that the uploaded data is correct.

I suggest to validate the uploaded files and/or implement custom filtering for coverage files via a simple flag.

**Comment by stevepeak** [comment]
I'm happy to do this and tried thus. Do you have suggestions to add:

- Matching patterns: https://github.com/codecov/codecov-bash/blob/ef22482d8e804579a9e855ca24b83bb7dd9bd72c/codecov#L632-L645
- File type detection: https://github.com/codecov/codecov-bash/blob/ef22482d8e804579a9e855ca24b83bb7dd9bd72c/codecov#L793-L797

Thank you!

**Comment by Flamefire** [comment]
I don't understand how this answers the issue. "Matching patterns" matches (wrongly in my case) a `*.lst` file than contains binary data. The "File type detection" seems to only check, if it is a file. I don't see where the files content is validated.
How is it even possible, that something was matched to the main.cpp file (in my case) out of (in respect to coverage data) garbage?

**Comment by stevepeak** [comment]
@Flamefire .... I'm not sure :( Whats a recent owner/repo@sha I can checkout

**Comment by Flamefire** [comment]
You can try this repo: https://github.com/Return-To-The-Roots/s25client
Currently our travis file is modified so it generates and collects coverage files itself, using the bash script directly uploaded the lst files (our data, not coverage related)
I think https://github.com/Return-To-The-Roots/s25client/commit/6dcf6371df88aab7d90b385c9a4f2c888ac4cf17 is the commit, that uploaded the wrong coverage data a while ago.

Side note: Current coverage report shows coverage data for our `src/test` folder although it is in the exclude list: https://github.com/Return-To-The-Roots/s25client/blob/0e21183be21900de164f702dafad972ac6b0cb44/codecov.yml -> https://codecov.io/gh/Return-To-The-Roots/s25client/tree/0e21183be21900de164f702dafad972ac6b0cb44/src

**Comment by stevepeak** [comment]
The yaml looks good. There were two issues I fixed about ~11hours ago **server side**

```diff
 coverage:
-    range: "65..95"
+    range: "65...95"
     ignore:
-        - contrib/*
+        - "contrib/*"
-        - src/test/*
+        - "src/test/*"
 comment:
     layout: "header, diff, changes"
```

You do not need to change these values, I've fixed the parser server side to accept them. The range was to sensitive on the 3 dots, now it accepts 2. And the default yaml parser fails on `- abc*` so we now regexp it to `- "abc*"` automatically.

**Comment by Flamefire** [comment]
Ok, so next coverage report should be correct in ignores.

Already any news on the wrong coverage for non-ignored files?

**Comment by stevepeak** [comment]
@Flamefire yea, I just checked that out. Your latest commit has `src/test` ignored right?

**Comment by stevepeak** [comment]
PS is this project starting from 0% coverage and your going to add tests and move that number up? If so: thats awesome :) I would love to keep tract of it and learn how your using Codecov in achieving high coverage.

**Comment by Flamefire** [comment]
Yes it seems it is now correct (ignore was added a few commits earlier: 7e1c9b099908a22c0745387540862a8e5cd1bfce)

Yes this is correct. History so far:
- We had no tests and hence no coverage
- Added `--coverage` flag and codecov bash -> Wrong coverage reports (no tests run, but binary lst files in the repo)
- Add `--coverage` only for debug and manually collect report files
- Add first tests -> Correct and incorrect coverage data

**Comment by Flamefire** [comment]
@stevepeak About the wrong coverage files: Is it possible to reset coverage data so the next Compile on master recreates it? Just to see if the incorrect coverage data still appears.

**Comment by stevepeak** [comment]
@Flamefire sorry, I don't quite understand your request. Correct my if I'm wrong in this example:

1. Travis job will report coverage with 3 builds
2. Codecov merges the 3 builds as expected
2. A travis rebuilds is triggered
3. Codecov starts from scratch

Currently we join all metrics and do not "delete" old builds

**Comment by Flamefire** [comment]
For us, this last joining is a problem and I don't see a reason for doing so. "Regular" example for how it should be:
- Commit 1 is build using 3 configurations
- Codecov merges these 3 reports which is the coverage report for Commit 1
- (Later) Commit 2 is build using 2 configurations (3rd configuration is removed so less files will be covered)
- Codecov merges (only) these 2 reports and present this as coverage for Commit 2

Especially I want no "left-over" data from Commit 1 in Commit 2. Why would you want that? This would lead me to the wrong assumption that the uncovered files in Commit 2 are still covered which is clearly not the case.

More importantly: In our case (and this issue) there were some files, that may or may not have looked like coverage data but are not. However codecov included them in the report and now there is now way of getting this old, wrong data out of the system.

I hope I'm not missing something here. Just in case I'm wrong and codecov **does** handle each commit on its own, then why is there data for [src/main.cpp](https://codecov.io/gh/Return-To-The-Roots/s25client/src/9c0c74345742fe7ffeaeff915d493a748b48ee59/src/main.cpp)? It is not in any report.

**Comment by gsauthof** [comment]
I can confirm, the file pattern rules to auto-detect coverage reports are way too greedy. In my tests, also  some unrelated `*.lst` files were wrongly identified as coverage reports. And even a cmake build file named `coverage.cmake` was uploaded and processed on codecov.io.

Perhaps you want to check some magic bytes (like [the fine free file command](http://www.darwinsys.com/file/)) instead of solely relying on filename patterns (at least on the server).

**Comment by Flamefire** [comment]
@stevepeak I got again (or still) the problem with incorrect attribution of coverage data for main.cpp which is never executed or included in the tests. See https://codecov.io/gh/Return-To-The-Roots/s25client/src/a93e9349cdc138bdb0b9ee77815535f13981e555/src/main.cpp in https://codecov.io/gh/Return-To-The-Roots/s25client/commit/a93e9349cdc138bdb0b9ee77815535f13981e555/builds

From what I see it attributes the coverage from `# path=#home#travis#build#Return-To-The-Roots#s25client#version#src#main.cpp.gcov` which is pretty much similar in name/path but certainly different than `#home#travis#build#Return-To-The-Roots#s25client#src#main.cpp.gcov` (which is not listed in the coverage report)
Proof for this assumption: Coverage report for version/src/main.cpp shows a first covered line at line 49 which is also the first line for which codecove reports coverage in src/main.cpp.

This seems to be a pretty simple but serious problem. Please fix this!

**Comment by stevepeak** [comment]
@gsauthof I can ignore the `*.cmake` files, but `*.lst` must remain because it's the coverage reports for D lang.

@Flamefire I'm seeing this in the raw uploads. So what your saying is that this file at `version/src/main.cpp` is being redirected to `src/main.cpp` correct? The quickest way to fix is adding "^src/main.cpp" to the ignore list of your codecov.yml. The problem is that this process of fixing paths is essential for other projects and I fear changing it would break path networks for other projects...

**Comment by Flamefire** [comment]
@stevepeak This is correct. The point is that I think it is a more general problem. Especially when using the long-path option of gcov it should be possible to correctly match the files and coverage reports. A problem like this can happen often: Files with the same name but in different folders. While it would be possible to add `main.cpp` to the ignore list in this case, I don't really trust the system that matches wrong files. What if I have more files with the same name in the reports?
So bottom line: full path is given. Multiple files are reported with full path. So it should be really really easy to match the full path to the correct files.
If you really expect problems with unmatched files, add an option to show unmatched files (e.g. coverage failure in CI) and an option to the yml file to manually match them or allow "fuzzy" matching as done right now, but not as a default for full path coverage reports.

**Comment by gsauthof** [comment]
@stevepeak - perhaps the lesson-learned from these reports is that just matching a filename pattern is too simplistic. Because there are too many false-positives - resulting in an fragile system/irritating user experience.

Perhaps a step forward would be to also look for magic bytes to detect the different coverage formats.

Ideally in the client, but at least at the server to reliably ignore any false-positives.

See also the [magic file specifications for some coverage formats I've contributed to the fine file command/libmagic](http://bugs.gw.com/view.php?id=549).

**Comment by stevepeak** [comment]
@gsauthof 

> just matching a filename pattern is too simplistic

I do agree, but it's working for the majority. I would love hear about better techniques that would improve this function.

> Magic bytes

I have not had experience with this at all. There are a dozen more coverage report formats not mentioned in the link provided too.. Can you share more about how you would implement this?

**Comment by Flamefire** [comment]
@stevepeak If you do some automatic matching (that is not even influencable by the user) why not use everything you have? For example you have the 2 coverage files in our project: `#home#travis#build#Return-To-The-Roots#s25client#version#src#main.cpp.gcov` and `#home#travis#build#Return-To-The-Roots#s25client#src#main.cpp.gcov` First of all you could detect, that a full path is specified: `/home/travis/build/Repo-Owner/Repo-Name/src/main.cpp.gcov` -> Hey we can even find this file in the repo. No guessing here. Same for `/home/travis/build/Repo-Owner/Repo-Name/version/src/main.cpp.gcov` although this is in a submodule. If you can detect it in the submodule you are done and can correctly match this. But I assume you can't. Then lets go further: We have a lot of files, that we can easily and uniquely match to the repo files. Namely all files in `/home/travis/build/Repo-Owner/Repo-Name/src/`. Hence we can assume `/home/travis/build/Repo-Owner/Repo-Name` to be the root path to the repo. Then we can match the previously unmatched files which leads to: `<rootPathToRepo>/version/src/main.cpp.gcov`. We will find `version` to be a valid subfolder (and even that it's a submodule) or at least, that `/version/src` is distinct from `/src` and hence we don't want it to match that. You already have some user defined path-replacing running. Hence you could show the user in codecov, that you have unmatched files that he may want to rematch or ignore. So in this case it should be possible to add a ignore rule to `version/*` and get rid of all the wrong stuff.
